src/gtc/passes/oir_optimizations/inlining.py

```diff
@@ -0,0 +1,130 @@
+# -*- coding: utf-8 -*-
+#
+# GTC Toolchain - GT4Py Project - GridTools Framework
+#
+# Copyright (c) 2014-2021, ETH Zurich
+# All rights reserved.
+#
+# This file is part of the GT4Py project and the GridTools framework.
+# GT4Py is free software: you can redistribute it and/or modify it under
+# the terms of the GNU General Public License as published by the
+# Free Software Foundation, either version 3 of the License, or any later
+# version. See the LICENSE.txt file at the top-level directory of this
+# distribution for a copy of the license or check <https://www.gnu.org/licenses/>.
+#
+# SPDX-License-Identifier: GPL-3.0-or-later
+
+import copy as cp
+from typing import Any, Dict, Optional
+
+from eve import NodeTranslator, NodeVisitor
+from gtc import oir
+
+
+class MaskCollector(NodeVisitor):
+    """Collects the boolean expressions definine mask statements that are boolean fields."""
```

Comment:
Thank you for the feedback @havogt, I have a question about this requirement. If we have code like the following,
```python
mask: Field[bool] = (conditional < 0.0)
if mask:
    conditional = input * 2.0
...
if mask:
    # More code...
```
In this case, the `mask` field can potentially contain incorrect results since `conditional` has changed in the first mask statement body.

However if the mask is inlined,
```python
if conditional < 0.0:
    conditional = input * 2.0
...
if conditional < 0.0:
    # More code...
```
There is no longer the possibility of stale mask data. Perhaps I do not fully understand how the masks work.

Reply:
It's explained here https://github.com/GridTools/concepts/blob/master/GTScript-Parallel-model.md#conditionals-on-field-expressions.

The condition of an `if` is just written to a field to avoid race conditions in the if-condition. That's the mask field. (see also statement at the end)

Assuming your examples are meant to be GTScript, then for your first example would be something like (in pseudo-OIR notation):
```
HorizontalExecution(mask=None) {
    mask = conditional < 0.0
}
HorizontalExecution(mask=None) {
    mask1 = mask
}
HorizontalExecution(mask=mask1) {
    conditional = input * 2.0
}
HorizontalExecution(mask=None) {
    mask2 = mask
}
HorizontalExecution(mask=mask2) {
    # More code...
}
```

If you example is meant to be pseudo-OIR, then there is nothing wrong with it. It cannot "potentially contain incorrect results" because semantic is well defined (but not the semantic that you had in mind).

The second example is a good example which could make sense as a regression test. Its semantic is again well-defined.

```
HorizontalExecution(mask=None) {
    mask1 = conditional < 0.0
}
HorizontalExecution(mask=mask1) {
    conditional = input * 2.0
}
HorizontalExecution(mask=None) {
    mask2 = conditional < 0.0
}
HorizontalExecution(mask=mask2) {
    # More code...
}
```

After having written all that and thinking about it again, I believe that mask field is an artifact from a previous iteration of the parallel model. So I guess the real fix is not to generate mask fields in the first place.